server/app/app.py

```diff
@@ -1,26 +1,189 @@
 import os
 import datetime
 
-from flask import Flask
+from flask import Flask, redirect, current_app, make_response, render_template
+from flask import Blueprint, request, send_from_directory
 from flask_caching import Cache
 from flask_compress import Compress
 from flask_cors import CORS
+from flask_restful import Api, Resource
 
-from server.app.rest_api.rest import get_api_resources
-from server.app.util.utils import Float32JSONEncoder
-from server.app.web import webapp
+from http import HTTPStatus
+
+import server.common.rest as common_rest
+from server.common.utils import path_join, Float32JSONEncoder
+from server.data_common.matrix_loader import MatrixDataLoader, MatrixDataType, MatrixDataCacheManager
+
+from functools import wraps
+
+webbp = Blueprint("webapp", "server.common.web", template_folder="templates")
+
+
+@webbp.route("/")
+def index():
+    if current_app.data is None:
+        return dataroot_index()
+
+    dataset_title = current_app.app_config.title
+    scripts = current_app.app_config.scripts
+    return render_template("index.html", datasetTitle=dataset_title, SCRIPTS=scripts)
+
+
+@webbp.route("/favicon.png")
+def favicon():
+    return send_from_directory(os.path.join(webbp.root_path, "static/img/"), "favicon.png")
+
+
+def get_data_adaptor(dataset):
+    config = current_app.app_config
+    location = path_join(config.dataroot, dataset)
+    matrix_data_loader = MatrixDataLoader(location)
+    matrix_data_loader.pre_load_validation()
+    data = matrix_data_loader.open(current_app.app_config)
+    return data
+
+
+def rest_get_data_adaptor(func):
+    @wraps(func)
+    def wrapped_function(self, dataset=None):
+        try:
+            if dataset is None:
+                # use the default dataset
+                data_adaptor = current_app.data
+                if data_adaptor is None:
+                    return make_response("Dataset must be supplied", HTTPStatus.BAD_REQUEST)
+                return func(self, data_adaptor)
+
+            config = current_app.app_config
+            location = path_join(config.dataroot, dataset)
```

Comment:
While this is currently safe, due to the way the route is specified (no slashes allowed), I suggest you add more safety to this statement in case `dataset` contains stuff like `..` or `../../../foo/`      Ideally it would be certain that the `location` is below the dataroot, and can't escape out using relative paths.

Reply:
Fixed.